src/validate-icon.c

```diff
@@ -36,55 +38,125 @@
 #define ICON_VALIDATOR_GROUP "Icon Validator"
 #define MAX_ICON_SIZE 512
 #define MAX_SVG_ICON_SIZE 4096
+#define BUFFER_SIZE 4096
+
+static void
+size_prepared_cb (GdkPixbufLoader *loader,
+                  int              width,
+                  int              height,
+                  gpointer         data)
+{
+  struct {
+    int width;
+    int height;
+    GdkPixbufFormat *format;
+  } *info = data;
+
+  g_return_if_fail (width > 0 && height > 0);
+
+  info->width = width;
+  info->height = height;
+  info->format = gdk_pixbuf_loader_get_format (loader);
+}
 
 static int
 validate_icon (const char *filename)
 {
-  GdkPixbufFormat *format;
   int max_size;
-  int width, height;
   g_autofree char *name = NULL;
   const char *allowed_formats[] = { "png", "jpeg", "svg", NULL };
   g_autoptr(GdkPixbuf) pixbuf = NULL;
   g_autoptr(GError) error = NULL;
   g_autoptr(GKeyFile) key_file = NULL;
   g_autofree char *key_file_data = NULL;
-
-  format = gdk_pixbuf_get_file_info (filename, &width, &height);
-  if (format == NULL)
+  g_autoptr(GdkPixbufLoader) loader = NULL;
+  g_autoptr(GFile) file = NULL;
+  g_autoptr(GInputStream) stream = NULL;
+  guchar buffer[BUFFER_SIZE];
+  gsize bytes_read = -1;
+  struct {
+    int width;
+    int height;
+    GdkPixbufFormat *format;
+  } info;
+
+  info.width = -1;
+  info.height = -1;
+  info.format = NULL;
+
+  loader = gdk_pixbuf_loader_new ();
+
+  file = g_file_new_for_commandline_arg (filename);
+  stream = G_INPUT_STREAM (g_file_read (file, NULL, &error));
+
+  if (!stream)
     {
-      g_printerr ("Format not recognized\n");
+      g_printerr ("Failed to load image: %s\n", error->message);
       return 1;
     }
 
-  if (width != height)
-    {
-      g_printerr ("Expected a square icon but got: %dx%d\n", width, height);
-      return 1;
-    }
+  g_signal_connect (loader, "size-prepared", G_CALLBACK (size_prepared_cb), &info);
 
-  name = gdk_pixbuf_format_get_name (format);
-  if (!g_strv_contains (allowed_formats, name))
+  while (bytes_read != 0)
     {
-      g_printerr ("Format %s not accepted\n", name);
-      return 1;
-    }
-
-  /* Sanity check for vector files */
-  max_size = g_str_equal (name, "svg") ? MAX_SVG_ICON_SIZE : MAX_ICON_SIZE;
+      bytes_read = g_input_stream_read (stream, buffer, BUFFER_SIZE, NULL, &error);
 
-  /* The icon is a square so we only need to check one side */
-  if (width > max_size)
-    {
-      g_printerr ("Image too large (%dx%d). Max. size %dx%d\n", width, height, max_size, max_size);
-      return 1;
-    }
+      if (bytes_read == -1)
+        {
+          g_printerr ("Failed to load image: %s\n", error->message);
+          gdk_pixbuf_loader_close (loader, NULL);
```

Comment:
I don't think you need to do this? Unreffing should be enough.

Reply:
No we need to close it before dropping the last reference https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-loader.c#L248

Reply:
No idea why that requirement exists.

Reply:
Potentially so you can get error reporting information out of `gdk_pixbuf_loader_close()`, and to make it easier to notice refcounting errors.

Reply:
Yikes, that's really not nice...

Reply:
If the reasoning is that (and I’m only guessing), it makes a lot of sense. Finalisation and shutdown are logically two separate things, and finalisation does not allow error reporting.

In addition, garbage collected runtimes may keep a strong reference to an object around for a long time after the code has actually stopped using it. If there’s no explicit shutdown step, the (potentially significant) resources in the pixbuf loader will remain around in memory for that unbounded period of time.

Reply:
We don't care about errors anymore at that point though and an explicit shutdown step could be provided while also doing it when the ref count drops to zero.

Anyway, that's the API we have...